### PR TITLE
[WIP] Working example of nginx reverse proxy with basic auth

### LINKED ISSUE #3211: Ability to connect to basic auth protected API.
## What is the motivation for adding / enhancing this feature?
If you set a basic auth on API, there is no way to connect to it from SSR VSF.

## What are the acceptance criteria 
* it's possible to connect to basic auth protected API

## Can you complete this feature request by yourself?
* [x] YES
* [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

* [x] This is a normal feature request. This should be available on https://test.storefrontcloud.io and then after tests this can be added to next Vue Storefront version. In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
* [ ] (**Pick this option only if you're sure**) This is an important improvement request for current Release Candidate version on https://next.storefrontcloud.io and should be placed in next RC version. In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
* [ ] (**Pick this option only if you're sure**) This is a critical improvement request for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version. In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

### PR DESCRIPTION
**Pull request by defudef**
### Related issues

closes #3211 

### Short description and why it's useful

TODO: I'll document this PR properly on Monday including changelog

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by pkarw** [comment]
Adding up to point 2) - if you really like to keep nginx here we can probably create kind of nginx config for production - including caching etc: https://docs.vuestorefront.io/guide/installation/production-setup.html#production-setup-bare-vps

then include this `docker-compose.reverse. ..` in the Production setup docs. However the authorization still: should be handled via middleware instead of `nginx`

**Comment by defudef** [comment]
@pkarw basically I've added nginx docker-compose file as a working example. In fact: you can but you don't have to use it and it's fully backward compatible. If it's Ok for you I can organize them more so people can use it.

In general, it is now possible to send requests from the server to the API with basic authentication enabled. However, I'll change it as you mentioned before :)

File: src/modules/authentication/server.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import passport from 'passport';
+import { GenerateStrategy, UnsupportedAuthenticationMethodError } from './common';
+
+export default (app, moduleConfig) => {
+  import(`./strategies/${moduleConfig.authMethod}.strategy.ts`)
```
Comment (defudef) [inline src/modules/authentication/server.ts]: I noticed several times, it's common to use dynamic imports depending on some variable. What do you think about this approach? In fact, if a developer wants to add another auth method (for example: OpenID), he need to just create file with convention above: `openid.strategy.ts` with a function of a type `GenerateStrategy`

Comment (pkarw) [reply src/modules/authentication/server.ts]: Isn't it adding the whole passport to the client bundle anyway? I belive it's not worth it if we can achieve the same logick anyway (I belive we can as you had this http password ttoken genreration code in a 1liner helper)

**Comment by pkarw** [comment]
OK, please apply the suggested changes anyway :)

File: src/modules/authentication/server.ts (hunk 1 above)
Comment (defudef) [reply src/modules/authentication/server.ts]: Is a `server.ts` file used in frontend as well? If so, then you're right and I'll remove these dependencies :)

Comment (pkarw) [reply src/modules/authentication/server.ts]: No, it shouldn't. So you're right we're safe here :)

**Comment by patzick** [comment]
@defudef what's the status of this PR? Is it still WIP? Are you willing to finish it for review? :)

**Comment by defudef** [comment]
@patzick @ResuBaka I'm going to finally finish it this week :)

**Pull request closed without merging**